cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,327 @@
+    }
+    finish();
+  }
+
+  private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    // Sanity check to confirm that the final distribution is brokerSet aware.
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+
+    for (Map.Entry<String, List<Partition>> partitionsByTopic : clusterModel.getPartitionsByTopic().entrySet()) {
+      if (!excludedTopics.contains(partitionsByTopic.getKey())) {
+        List<Partition> partitions = partitionsByTopic.getValue();
+        Set<Broker> allBrokersForTopic = partitions.stream()
+                                                   .map(partition -> partition.partitionBrokers())
+                                                   .flatMap(brokers -> brokers.stream())
+                                                   .collect(Collectors.toSet());
+        // Checks if a topic's brokers do not all live in a single brokerSet
+        if (_brokersByBrokerSet.values().stream().noneMatch(brokerSetBrokers -> brokerSetBrokers.containsAll(allBrokersForTopic))) {
+          throw new OptimizationFailureException(
+              String.format("[%s] Topic %s is not brokerSet-aware. brokers (%s).", name(), partitionsByTopic.getKey(), allBrokersForTopic));
+        }
+      }
+    }
+  }
+
+  /**
+   * brokerSet-awareness violations can be resolved with replica movements.
+   *
+   * @param broker              Broker to be balanced.
+   * @param clusterModel        The state of the cluster.
+   * @param optimizedGoals      Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker, ClusterModel clusterModel, Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+
+    String currentBrokerSetId = _brokerSetResolutionHelper.brokerSetId(broker);
+
+    for (Replica replica : broker.trackedSortedReplicas(replicaSortName(this, false, false)).sortedReplicas(true)) {
+      String eligibleBrokerSetIdForReplica =
+          _replicaToBrokerSetMappingPolicy.brokerSetIdForReplica(replica, clusterModel, _brokerSetResolutionHelper);
+
+      // If the re-balancing action is performed within the broker set boundary, we are good
+      if (broker.isAlive() && eligibleBrokerSetIdForReplica.equals(currentBrokerSetId)) {
+        continue;
+      }
+      // If the brokerSet awareness condition is violated. Move replica to an eligible broker
+      Set<Broker> eligibleBrokers = _brokersByBrokerSet.get(eligibleBrokerSetIdForReplica)
+                                                       .stream()
+                                                       .filter(b -> clusterModel.aliveBrokers().contains(b))
```

Comment:
I tried filtering this on the BrokerSetResolverHelper but the ExcludedBrokersForLeaderShipTest fails since it marks excluded brokers as dead and clusterModel.aliveBrokers() do not contain the excluded broker for leadership.

Reply:
2 questions:
1. Could you point me to where it marks excluded brokers as dead?
2. Even if clusterModel.aliveBrokers() does not contain excluded broker, the test should succeed? E.g. if you have enough alive brokers, I don't see the reason why this goal can fail.